lib/queryContextCache.js

```diff
@@ -0,0 +1,391 @@
+};
+
+/**
+ * @param data: the QueryContext Object serialized as a JSON format string
+ */
+QueryContextCache.prototype.deserializeQueryContext = function (data) {
+  Logger.getInstance().debug(`deserializeQueryContext() called: data from server: ${JSON.stringify(data)}`);
+  if (!data||JSON.stringify(data)==='{}'||data.entries === null) {
+
+    // Clear the cache
+    this.clearCache();
+    Logger.getInstance().debug('deserializeQueryContext() returns');
+    this.logCacheEntries();
+    return;
+  }
+
+  try {
+    // Deserialize the entries. The first entry with priority is the main entry. On JDBC side,
+    // we save all entries into one list to simplify the logic. An example JSON is:
+    // {
+    //   "entries": [
+    //    {
+    //     "id": 0,
+    //     "readtimestamp": 123456789,
+    //     "priority": 0,
+    //     "context": "base64 encoded context"
+    //    },
+    //     {
+    //       "id": 1,
+    //       "readtimestamp": 123456789,
+    //       "priority": 1,
+    //       "context": "base64 encoded context"
+    //     },
+    //     {
+    //       "id": 2,
+    //       "readtimestamp": 123456789,
+    //       "priority": 2,
+    //       "context": "base64 encoded context"
+    //     }
+    //   ]
+
+    const entries = data.entries;
+    if (entries !== null && Array.isArray(entries)) {
+      for (const entryNode of entries) {
+        const entry = this.deserializeQueryContextElement(entryNode);
+        if (entry != null) {
+          this.merge(entry.getId(), entry.getReadTimestamp(), entry.getPriority(), entry.getContext());
+        } else {
+          Logger.getInstance().warn(
+            'deserializeQueryContextJson: deserializeQueryContextElement meets mismatch field type. Clear the QueryContextCache.');
+          this.clearCache();
+          return;
+        }
+      }
+      // after merging all entries, sync the internal priority map to priority map. Because of
+      // priority swicth from GS side,
+      // there could be priority key conflict if we directly operating on the priorityMap during
+      // a round of merge.
+      this.syncPriorityMap();
+    }
+  } catch (e) {
+    Logger.getInstance().debug(`deserializeQueryContextJson: Exception = ${e.getMessage}`, );
+
+    // Not rethrowing. clear the cache as incomplete merge can lead to unexpected behavior.
+    this.clearCache();
+  }
+
+  // After merging all entries, truncate to capacity
+  this.checkCacheCapacity();
+
+  // Log existing cache entries
+  this.logCacheEntries();
+}; // Synchronized
+
+QueryContextCache.prototype.deserializeQueryContextElement = function (node) {
+  const entry = new QueryContextElement();
+  const idNode = node.id;
+  if (typeof idNode === 'number') {
+    entry.setId(idNode);
+  } else {
+    Logger.getInstance().warn('deserializeQueryContextElement: `id` field is not Number type');
+    return null;
+  }
+
+  const timestampNode = node.timestamp;
+  if (typeof timestampNode === 'number') {
```

Comment:
do we need such checks on Snowflake response?

Reply:
let's make it more in the Javascript style

Reply:
I removed unnecessary checkings (id, timestamp, priority).